### ISSUE TITLE
P0: the SVD adjoint kernel is wrong — AD gradients disagree with finite differences in DIRECTION (cos = -0.57), affecting every SVD-based AD path

### ISSUE THREAD

**Issue by yingjerkao** [tenax-lab/tenax#750]
Found while investigating #687. This is independent of, and much larger than, that issue.

## Claim

`_svd_sector_backward` in `src/tenax/algorithms/_ad_primitives.py` computes its off-diagonal (F-matrix) contribution at **−½ × the correct value**. The resulting gradient does not merely lose precision — it points in a materially **wrong direction**.

## Evidence: AD vs finite differences of its own forward

Gauge-free by construction — FD is taken on exactly the function being differentiated, so every sign/gauge convention cancels. 6×6 real matrix, x64, `loss = sum(W*U) + sum(Z*Vh) + sum(s)`:

| case | shipped kernel | with the fix below |
|---|---|---|
| **full rank** (k=n; terms 4/5 vanish) | rel **1.4126e+00**, cos **−0.570746546** | rel **2.7317e-09**, cos **1.000000000** |
| truncated (k=3) | rel 1.4004e+00, cos −0.533174323 | rel 6.5879e-02, cos 0.997829417 |

A negative cosine means the reported gradient has a component pointing *away* from the true one.

Cross-check against JAX's own SVD VJP, projecting out the shared diagonal (`ds`) part:

```
off-diagonal projection  <tenax, jax> / |jax|^2 = 0.500000000
```

Exactly one half, to nine digits.

## Root cause — two lines

```python
# src/tenax/algorithms/_ad_primitives.py
diff = s2[:, None] - s2[None, :]          # line 211: F_ij ~ 1/(s_i^2 - s_j^2); sign flipped
UtdU_anti = 0.5 * (UtdU - UtdU.conj().T)  # line 219: standard adjoint uses the full (J - J^H)
VtdV_anti = 0.5 * (VtdV - VtdV.conj().T)  # line 220
```

The standard SVD adjoint needs `F_ij = 1/(s_j² − s_i²)` and the *unhalved* antisymmetric part. Combined, terms 2 and 3 (lines 239, 242) come out at −0.5×.

**Precedent in this very file:** the sibling `regularized_eigh` backward at `_ad_primitives.py:545` carries the comment *"The previous formulation used `w_i − w_j` which flipped the sign of the backward (issue #316)"*. `eigh` was fixed for exactly this class of error. The SVD kernel never was.

## Proposed fix

```python
diff = s2[None, :] - s2[:, None]
UtdU_anti = UtdU - UtdU.conj().T
VtdV_anti = VtdV - VtdV.conj().T
```

Leave `keep_mask`, the `s_inv` sanitisation, and terms 4/5 alone.

## Why no test caught it

**107 core tests pass identically with and without the fix** (`test_ad_primitives_rank_aware`, `test_regularized_svd`, `test_linalg`, `test_ctm_projector`, `test_split_ctm_2site_ad`, `test_ad_utils`, `test_projector_backward_dispatch`, `test_ctm_root_implicit_symmetric`, `test_ctm_root_implicit_asym`, `test_split_ctm_2site_symmetric`).

No existing test compares an SVD-path gradient to a trustworthy reference. They assert finiteness, direction (`cos > 0.99`), determinism, or `implicit == explicit` — and that last one **cancels the bug**, since both sides share the same primitive. That is the gap to close: a direct AD-vs-FD gate on the primitive itself.

## The "Wirtinger gap" in the test suite is this bug

`tests/test_split_ctm_2site_ad.py:8,162,342`, `test_split_ctm_fuse_flag.py:175`, and `test_split_ctm_2site_symmetric.py:267` all attribute a dense AD-vs-FD magnitude mismatch "including sign flips" to a Wirtinger/holomorphy gap. **Everything in that computation is real**, so Wirtinger cannot apply. Measured on the dense implicit split energy:

```
shipped: AD-vs-FD cos=0.849233800  rel=5.3895e-01   (AD +0.00532561 vs FD -0.00647834 -- sign flip)
fixed:   AD-vs-FD cos=1.000000000  rel=6.1598e-06
```

Those docstrings should be corrected, not just the code.

## Blast radius — every SVD-based AD path

`_svd_sector_backward` is the backward for:

- `truncated_svd_ad` → `_ctm_compiled_moves.py:87` (fused/compiled CTM projector), `_ctm_projector.py:986`, `linalg.py:879` (every block-sparse SVD under tracing, i.e. **all symmetric AD**)
- `truncated_svd_symmetric_ad` → the dense `_svd_split_edge_tensor`
- `truncated_svd_ad_vh_only` (exported via `ad_utils`)
- `regularized_svd` → `_ctm_projector.py:1097`

So a fix moves **every** iPEPS CTM-AD gradient — dense and symmetric, fused and split, 1×1 and 2×2.

## Before landing

**The production-energy impact is NOT established.** Gradient direction being wrong does not automatically mean converged energies are wrong — a descent method can still reach a minimum along a distorted direction, more slowly. But it does mean every published optimization trajectory, convergence rate, and gradient-norm claim needs re-checking. Measure end-to-end `optimize_gs_ad` energies with and without the fix before merging, and run the full `-m core` plus the slow/algorithm tiers.

**Honest caveat on my own numbers:** the full-rank case is machine-exact after the fix (2.7e-9). My truncated k=3 probe still shows 6.6e-2 — I did not chase whether that is FD noise across the truncation boundary or a residual defect in terms 4/5. Worth resolving as part of the fix.

## Suggested sequencing

1. Land the two-line fix behind a real AD-vs-FD test on the primitive (full rank, then truncated).
2. Re-measure production energies; expect trajectories to change.
3. Correct the "Wirtinger gap" docstrings.
4. Tighten the gates that were loosened to accommodate this: `test_split_ctm_2site_symmetric.py:333` (`5e-2`) and `test_split_ctm_2site_ad.py:339` (`cos > 0.99`).

Related: #316 (the same error class, fixed in `eigh`), #687 (surfaced this; keep open for its residual ~2.5e-4 traced-vs-eager symmetric floor).

> 🤖 **AI-generated issue** — written by Claude Code, posted by @yingjerkao.

**Comment by yingjerkao** [comment]
## Correction to the blast-radius claim in this issue

The issue states the fix "moves **every** iPEPS CTM-AD gradient — dense and symmetric, fused and split, 1×1 and 2×2." **The 2×2 part is wrong**, and 2×2 is the default `gs_recipe`.

Measured by wrapping `_svd_sector_backward` in a counter and running `optimize_gs_ad` to completion (D=2 Heisenberg, `gs_num_steps=2`, χ=4):

| config | `_svd_sector_backward` calls |
|---|---|
| `gs_recipe="2x2"`, fused (**the default**) | **0** |
| `gs_recipe="1x1"`, fused | 7 |
| `gs_recipe="1x1"`, split | 94 |

The 2×2 projector does not use these primitives at all: `_ctm_tensor_projector_2x2.py` calls `_dense_svd` through its own local `_gauge_fixed_svd`, so the backward is **JAX's own** SVD VJP — which is correct. That is confirmed end-to-end: with `gs_recipe="2x2"`, optimized energies are **bit-identical** before and after the fix.

```
chi= 8   before -0.660183823708   after -0.660183823708
chi=16   before -0.660093885590   after -0.660093885590
```

### Corrected blast radius

Affected:
- **All symmetric / block-sparse SVD under AD** — `linalg.py:691` and `:879` call `truncated_svd_ad` per charge sector. This is the widest exposure and is independent of CTM recipe.
- **Dense CTM `1x1`**, fused (`_ctm_compiled_moves.py:87`) and split.
- `regularized_svd` → `_ctm_projector.py:1097` (reduced-corner / QR projector path).
- `truncated_svd_ad_vh_only`, exported via `ad_utils`.

Not affected:
- **Dense CTM `2x2`** — the default recipe.

### What that means for historical results

`gs_recipe="1x1"` energies do move. Same driver, D=2, 20 optimizer steps:

| config | before | after |
|---|---|---|
| `1x1` fused, χ=8 | -0.071644964555 | -0.087911674853 |

Those numbers are not physically meaningful either way — `1x1` also carries the #726 rank-1 corner collapse — so this is evidence that the gradient *changed*, not that it improved. Any study run on `1x1` was subject to **both** defects at once.

The practical upshot is narrower than the issue implied but still real: default-recipe dense iPEPS runs are unaffected, while every symmetric/block-sparse AD gradient was wrong.

> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.
